rust/hyperlane-base/src/contract_sync/cursor.rs

```diff
@@ -1,90 +1,348 @@
+    /// The next nonce that the cursor is looking for.
+    next_nonce: u32,
+}
+
+impl MessageSyncCursor {
+    async fn retrieve_dispatched_block_number(&self, nonce: u32) -> Option<u32> {
+        if let Ok(Some(block_number)) = self.db.retrieve_dispatched_block_number(nonce).await {
+            Some(u32::try_from(block_number).unwrap())
+        } else {
+            None
+        }
+    }
+
+    async fn update(
+        &mut self,
+        logs: Vec<(HyperlaneMessage, LogMeta)>,
+        prev_nonce: u32,
+    ) -> eyre::Result<()> {
+        // If we found messages, but did *not* find the message we were looking for,
+        // we need to rewind to the block at which we found the last message.
+        if !logs.is_empty() && !logs.iter().any(|m| m.0.nonce == self.next_nonce) {
+            // If the previous nonce has been synced, rewind to the block number
+            // at which it was dispatched. Otherwise, rewind all the way back to the start block.
+            if let Some(block_number) = self.retrieve_dispatched_block_number(prev_nonce).await {
+                self.next_block = block_number;
+            } else {
+                self.next_block = self.start_block;
+            }
+            Ok(())
+        } else {
+            Ok(())
+        }
+    }
+}
+
+/// A MessageSyncCursor that syncs forwards in perpetuity.
+#[derive(new)]
+pub(crate) struct ForwardMessageSyncCursor(MessageSyncCursor);
+
+impl ForwardMessageSyncCursor {
+    async fn get_next_range(&mut self) -> ChainResult<Option<(u32, u32, Duration)>> {
+        // Check if any new messages have been inserted into the DB,
+        // and update the cursor accordingly.
+        while let Some(block_number) = self
+            .0
+            .retrieve_dispatched_block_number(self.0.next_nonce)
+            .await
+        {
+            self.0.next_nonce += 1;
+            self.0.next_block = block_number;
+        }
+
+        let (mailbox_count, tip) = self.0.indexer.fetch_count_at_tip().await?;
+        let cursor_count = self.0.next_nonce;
+
+        let cmp = cursor_count.cmp(&mailbox_count);
+        match cmp {
+            Ordering::Equal => {
+                // We are synced up to the latest nonce so we don't need to index anything.
+                // We update our next block number accordingly.
+                self.0.next_block = tip;
+                Ok(None)
+            }
+            Ordering::Less => {
+                // The cursor is behind the mailbox, so we need to index some blocks.
+                // We attempt to index a range of blocks that is as large as possible.
+                let from = self.0.next_block;
+                let to = u32::min(tip, from + self.0.chunk_size);
+                self.0.next_block = to + 1;
+                Ok(Some((from, to, Duration::from_secs(0))))
+            }
+            Ordering::Greater => {
+                panic!("Cursor is ahead of mailbox, this should never happen.");
+            }
+        }
```

Comment:
```suggestion
        if cursor_count == mailbox_count {
            // We are synced up to the latest nonce so we don't need to index anything.
            // We update our next block number accordingly.
            self.0.next_block = tip;
            Ok(None)
        } else if cursor_count < mailbox_count {
            // The cursor is behind the mailbox, so we need to index some blocks.
            // We attempt to index a range of blocks that is as large as possible.
            let from = self.0.next_block;
            let to = u32::min(tip, from + self.0.chunk_size);
            self.0.next_block = to + 1;
            Ok(Some((from, to, Duration::from_secs(0))))
        } else {
            panic!("Cursor is ahead of mailbox, this should never happen.");
        }
```

Reply:
Clippy prefers the match statement

Reply:
What!? That looks so odd to me.

Funny, seems a few people agree with me on this:
https://github.com/rust-lang/rust-clippy/issues/4531